### ISSUE TITLE
Reproducibility issue

### ISSUE THREAD

**Issue by bfdyoy** [dotnet/TorchSharp#898]
I just created two binary classification models and after several runs, the results are always different. Things that I used so far:
-  torch.random.manual_seed()
-  setting the seed for the shuffler used when loading the data
-  turning off the shuffling 
But none of these proved to be the problem. The dataset that's loaded is already splitted (2 json files for each model, 1 for train and 1 for test).

**Comment by NiklasGustafsson** [comment]
Do you set the seed on CUDA, too (assuming that you use a GPU)?

**Comment by bfdyoy** [comment]
I didn't do any changes to the device, I don't even have a GPU chip and from what I saw, all the tensor were on CPU. Do I need to set the device to CPU or does it fallback on CPU ? I mean my GPU chip is an  Intel HD Graphics 630 and from my what I know it doesn't support CUDA.

**Comment by NiklasGustafsson** [comment]
No, the default is CPU, so there's no need to do anything.

**Comment by bfdyoy** [comment]
This is the code I used to create the NN. Below you can also find the training function.

`_dense1 = Linear(inputSize: BERTSize * 2, outputSize: BERTSize);
_dense2 = Linear(inputSize: BERTSize, outputSize: BERTSize / 4);
_dense3 = Linear(inputSize: BERTSize / 4, outputSize: BERTSize / 8);
_output = Linear(inputSize: BERTSize / 8, outputSize: 1);

Backbone = new List<Linear> { _dense1, _dense2, _dense3 };

_leakyRelu = LeakyReLU(negative_slope: 0.1f);
_sigmoid = Sigmoid();

RegisterComponents();
ModelVersion = modelVersion;`

`torch.enable_grad();
            model.train();
            var loss = new BCELoss();
            var scheduler = lr_scheduler.StepLR(optimizer, step_size: 5, gamma: 0.95);
            
            var trainDataLoader = new data.DataLoader(trainDataset, batchSize: batchSize, seed: Seed, shuffle:false);
            var trainDataEnumerator = trainDataLoader.GetEnumerator();

            foreach (var epoch in Enumerable.Range(0, numberEpochs))
            {
                var epochF1 = 0.0f;
                var epochLoss = 0.0f;
                using (var _ = NewDisposeScope())
                {
                    foreach (var batch in Enumerable.Range(0, (int)trainDataLoader.Count))
                    {
                        trainDataEnumerator.MoveNext();

                        var firstInputs = trainDataEnumerator.Current["firstInput"];
                        var secondInputs = trainDataEnumerator.Current["secondInput"];
                        var label = trainDataEnumerator.Current["label"];

                        model.zero_grad();
                        var outputs = model.forward(firstInputs, secondInputs);

                        var batchLoss = loss.forward(outputs, label.reshape(-1,1));
                        epochLoss += batchLoss.ToSingle();

                        var predictedOutput = ExtractDataFromTensor(outputs);
                        var trueLabels = ExtractDataFromTensor(label);

                        var (truePositives, falsePositives, _, falseNegatives) = ConfusionMatrix.CalculateConfusionMatrix(trueLabels, predictedOutput);

                        var batchF1 = ClassificationMetrics.ComputeF1Score(truePositives, falsePositives, falseNegatives);
                        epochF1 += batchF1;

                        optimizer.zero_grad();
                        batchLoss.backward();
                        optimizer.step();
                    }
                }
                Console.WriteLine($"Loss at epoch {epoch} is {epochLoss / trainDataLoader.Count}");
                Console.WriteLine($"F1 score at epoch {epoch} is {epochF1 / trainDataLoader.Count}");
                Console.WriteLine("---------------------------------------\n");

                scheduler.step();
                trainDataEnumerator.Reset();
            }`

**Comment by NiklasGustafsson** [comment]
What is it that is different between runs? The F1 score?

**Comment by bfdyoy** [comment]
Here are the results for 3 runs: 
run#1
model_1 test F1: 0.8079
model_2 test F1: 0.6805
run#2
model_1 F1: 0.8114
model_2 test F1: 0.6805
run#3
model_1 F1: 0.7428
model_2 test F1:0.6753

**Comment by bfdyoy** [comment]
Seems that the problems was actually on my part, I did the manual_seed oen step after creating the models so it's obvious why it didn't work. Thank you !

**Comment by NiklasGustafsson** [comment]
Yes, that will do it... :-) Glad you found it!
